doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,615 @@
+        page = cursor.get()
+        if not page:
+            return related
+        related += page
+
+results = extract_single(cursors[idx])
+```
+
+```python
+# Skip to nth page before extracting remaining results.
+
+def extract_all_after_n(page_num, page_size, cursors):
+    for _ in range(page_num):
+        for cursor in cursors:
+            cursor.next(page_size)
+
+    return extract_all(cursors)
+```
+
+#### Pros
+
+- Individual cursors can be incremented independently, potentially
+  reducing wasted queries.
+
+### Non-batched cursor object
+
+Host can only query a single entity-relationship pair.
+
+We assume a separate batched but non-paged signature, so that batching
+of related entity queries is still possible, as long as paging is not
+required at the same time.
+
+```python
+def relatedEntities(
+        input: EntityReference, relationship) -> AbstractPage: ...
+```
+
+```python
+# Usage:
+
+input = EntityReference(...)
+relationship = SomeRelationship.traitsData()
+
+# Get intermediate cursor object.
+cursor = relatedEntities(input, relationship)
+```
+
+```python
+# Extract all results.
+
+def extract(cursor):
+    related = []
+    while True:
+        cursor.next(100)
+        # Perhaps yield here to allow UI updates etc...
+        page = cursor.get()
+        if not page:
+            return related
+        related += page
+```
+
+#### Pros
+
+- Simple direct interface for the host application.
+- Manager implementation can typically be a lightweight wrapper around
+  common HTTP API pagination patterns.
+
+### Whole-batch cursor object
+
+Host supplies a list of entity references and is given a cursor object.
+Pages are then fetched for every element in the batch at the same time,
+by operating on the batch-aware cursor object.
+
+```python
+class AbstractPage:
+
+    def get(self) -> List[List[EntityReferences]]:
+        # Current page of refs for each element of input batch.
+        ...
+
+    def next(self):
+        ...
+
+    def remaining(self) -> Union[List[int], None]:
+        # None means "unsupported".
+        ...
+
+def relatedEntities(
+        batch: List[EntityReference], relationship) -> AbstractPage: ...
+```
+```python
+# Usage:
+
+inputs = [EntityReference(...), ...]
+relationship = SomeRelationship.traitsData()
+
+# Get intermediate cursor object.
+cursor = relatedEntities(inputs, relationship)
+```
+
+```python
+# Extract all results.
+
+def extract_all(inputs, cursor):
+    relateds = [[] for _ in inputs]
+
+    has_more_pages = True
+    while has_more_pages:
+        # Perhaps yield here to allow UI updates etc...
+        has_more_pages = False
+        cursor.next(100)
+        pages = cursor.get()
+        for page, related in zip(pages, relateds):
+            if page:
+                has_more_pages = True
+                related += page
+
+    return relateds
+```
+
+#### Cons
+
+- Additional metadata in the object is limited to whole-batch data.
+
+## Outcome
```

Comment:
[fix] Great work on the restructure. Can we move this before the appendix with out conclusions to date

- no batch paging
- cursor or token

Reply:
Have we definitively concluded no batch paging?  I had spent some time trying (mostly failing) to sketch how that would look on the manager side, but if we're confident we've rejected batch paging then I can skip that.

Reply:
We were strongly advised to keep the two things separate if we can. The only noted 'con' would be that it precludes complex UIs requesting their data in one go. It was noted though that we have no concrete use cases for that yet though, and can always add it later once we have some real world scenarios.

Reply:
Added an(other) intermediate Outcome section in 3854ec13